### PR TITLE
Adopt JEI search tree

### PR DESCRIPTION
**Pull request by LasmGratel**
A build of this version of NEI can be found [here](https://github.com/LasmGratel/NotEnoughItems/releases/tag/v2.6.1-beta)

List of changed/added features:
- IngredientListElement\<T\>, but the only ingredient type now is ItemStack
  It is possible in 1.7.10 because Mods register their fluid as a fake item, like GT6's gregapi:gt.display.fluid, or just buckets
- PrefixInfo, ElementSearch
  IngredientFilter are not added because ItemList contains it's functionality, ItemFilter is preserved for compatibility reasons but marked for deprecated and has no actual use
- New settings in NEIClientConfig
  Mod Name, Tooltip, ResourceId, OreDict, CreativeTab search mode changing like in JEI. But requires restart and needs to be localized.
- Shadow fastutil-8.5.8
  For best performance results, why not?
- Compatibility with NotEnoughCharacters
  Won't crash, but won't work either. I backported Hech and a build can be found [here](https://github.com/LasmGratel/JustEnoughCharacters/releases/tag/v1.5.2).
- Compatibility with NEI Addons, NEI Integrations
  Works like a charm
- Compatibility with others
  Integration mod that adds ISearchProvider or ItemFilter will have no actual effect, e.g. old NotEnoughCharacters.

Pros/Cons:
\+ Almost instant searching
\+ Configurable Search Modes and `|` splitable tokens for multiple searching
\+ Less CPU Usage
\- Possible compatibility issue
\- Higher memory usage(approx. 180MB for a large-scale modpack and even higher when using NEch)
\- Super slow first-run item list build time
\- Increases mod size from 800KB to 2.4MB

Tested in latest GTNH and my own Reika modpack.

### CODE REVIEW

File: src/main/java/codechicken/nei/ItemHistoryPanel.java [hunk 1]
```diff
@@ -39,8 +39,8 @@ public void update() {
 
     public void addItem(ItemStack stack) {
         if (stack != null) {
-            ItemStack is = StackInfo.loadFromNBT(StackInfo.itemStackToNBT(stack), 0);
-
+            ItemStack is = stack.copy();
```
Comment (slprime) [inline src/main/java/codechicken/nei/ItemHistoryPanel.java:42]: StackInfo is used here specifically to reset the fluid amount

remove this
image

**Comment by slprime** [comment]
Does not work presets & subsets. players use both things

image

image

**Comment by slprime** [comment]
I'm not sure that for the sake of searching it's worth inflating the mod so much. It's worth finding a more elegant solution

**Comment by LasmGratel** [comment]
Yeah, maybe just leave it for a while

**Comment by mitchej123** [comment]
> I'm not sure that for the sake of searching it's worth inflating the mod so much. It's worth finding a more elegant solution

Depend on gtnhlib which ships fastutils instead of shadowing it here

**Comment by Dream-Master** [comment]
Let this pr open so we can look if we find a solution or can it close later.

**Comment by slprime** [comment]
I don’t know how to show this more correctly, but the new algorithm remove filter visually slower

Details:
1. I use history so that the line is entered full and does not depend on the input speed
1. you need to compare the time between changing the text in the input and changing the panel on the right
1. video speed 0.25
1. gtnh version 2.6.0

https://github.com/GTNewHorizons/NotEnoughItems/assets/31038811/cd44c028-3269-496d-8468-dda8e1e22e85

https://github.com/GTNewHorizons/NotEnoughItems/assets/31038811/1137c29d-963c-4fed-9c5f-f985206665bc

**Comment by slprime** [comment]
I'm still not sure about the wisdom of these changes.

> Configurable Search Modes and | splitable tokens for multiple searching

can be implemented in a much simpler way

**Comment by Caedis** [comment]
Have you taken a look at https://github.com/CleanroomMC/HadEnoughItems/tree/cleanroom/src/main/java/mezz/jei/search instead of base jei ?

**Comment by slprime** [comment]
I compared NEI in 2.6.0 and the implementation in this PR

**Comment by Ethryan** [comment]
Needs spotless

**Comment by Dream-Master** [comment]
i am on it

**Comment by Dream-Master** [comment]
SLPrime — Today at 5:12 PM
this PR not need. if you approve https://github.com/GTNewHorizons/NotEnoughItems/pull/485 
DreamMasterXXL — Today at 5:12 PM
ah ok good to know. so i can close it ?
SLPrime — Today at 5:13 PM
yes

**Pull request closed without merging**